plugins/outputs/sematext/processors/metainfo.go

```diff
@@ -0,0 +1,247 @@
+package processors
+
+import (
+	"fmt"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/outputs/sematext/sender"
+	"math/rand"
+	"sync"
+	"time"
+)
+
+const (
+	metainfoRetryIntervalSeconds = 60
+	maxMetainfoRetries           = 10
+)
+
+// MetricMetainfo contains metainfo about a single metric
+type MetricMetainfo struct {
+	token       string
+	name        string
+	namespace   string
+	semType     SematextMetricType
+	numericType NumericType
+	label       string
+	description string
+	host        string
+}
+
+// SematextMetricType is an enumeration of metric types expected by Sematext backend
+type SematextMetricType int
+
+// Possible values for the ValueType enum.
+const (
+	_ SematextMetricType = iota
+	Counter
+	Gauge
+)
+
+func (s SematextMetricType) String() string {
+	return [...]string{"", "Counter", "Gauge"}[s]
+}
+
+// NumericType represents metric's data type
+type NumericType int
+
+const (
+	UnsupportedNumericType NumericType = iota
+	Long
+	Double
+	Bool
+)
+
+func (s NumericType) String() string {
+	return [...]string{"", "Long", "Double", "Bool"}[s]
+}
+
+// Metainfo is a processor that extracts metainfo from telegraf metrics and sends it to Sematext backend
+type Metainfo struct {
+	log         telegraf.Logger
+	token       string
+	sentMetrics map[string]*MetricMetainfo
+	lock        sync.Mutex
+	sendChannel chan bool
+	serializer  *MetainfoSerializer
+
+	metainfoURL  string
+	senderConfig *sender.Config
+	sender       *sender.Sender
+}
+
+// NewMetainfo creates a new Metainfo processor
+func NewMetainfo(log telegraf.Logger, token string, receiverURL string, senderConfig *sender.Config) BatchProcessor {
+	sentMetricsMap := make(map[string]*MetricMetainfo)
+	return &Metainfo{
+		log:          log,
+		token:        token,
+		sentMetrics:  sentMetricsMap,
+		sendChannel:  make(chan bool, 1),
+		metainfoURL:  receiverURL + "/write?db=metainfo",
+		senderConfig: senderConfig,
+		sender:       sender.NewSender(senderConfig),
+		serializer:   NewMetainfoSerializer(log),
+	}
+}
+
+// Process contains core logic of Metainfo processor
+func (m *Metainfo) Process(metrics []telegraf.Metric) ([]telegraf.Metric, error) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	newMetrics := make(map[string]*MetricMetainfo)
+
+	for _, metric := range metrics {
+		for _, field := range metric.FieldList() {
+			mInfo, mKey := processMetric(m.token, metric, field, m.sentMetrics)
+			if mInfo != nil {
+				newMetrics[mKey] = mInfo
+			}
+		}
+	}
+
+	if len(newMetrics) > 0 {
+		go m.sendMetainfo(newMetrics)
```

Comment:
This is completely unexpected as we are in a _PROCESSOR_ and now you start sending things. I think you should collect the metadata here and return it to the caller. Then send the stuff from the main loop. This would also avoid a lot of locking, channeling, etc.